### PR TITLE
Update and improve conformance of S3 API

### PR DESCRIPTION
**Pull request by ZacBlanco**
### What changes are proposed in this pull request?

Improve conformance of the S3 API to the specification defined by AWS.

### Why are the changes needed?

The AWS CLI does not work for many standard operations (e.g. listing buckets, setting max number of return keys, copying between buckets)

### Does this PR introduce any user facing changes?

This affects some of the values from the S3 API to be in-line with more of the S3 REST API specification. It can potentially break backwards compatibility with previous versions. 

### PR Summary

This PR is a a serious of small changes and improvements to the S3
API which improves the conformance to the specification defined
by the AWS Rest API reference at
https://docs.aws.amazon.com/AmazonS3/latest/API/Type_API_Reference.html

A short summary of the changes:

- Implement the CopyObjectResult type for bucket-bucket transfers
- Make the ListAllMyBucketsResult conform to the spec. Added
creationDate and proper timestamps to results
- ListBucketOptions now includes the delimiter and encoding type
parameters
- the listBucketResult uses the CommonPrefix type and now includes
the delimiter and encoding type. Additionally, make sure keyCount
and truncation takes into account the common prefixes
-  Update the S3RestServiceHandler to use Java 8 lambda language
features
- Update credentials to use the latest amazon authorization format
- A potentially breaking change which causes a ListAllMyBuckets result
to only return folders

### CODE REVIEW

File: core/server/proxy/src/main/java/alluxio/proxy/s3/ListAllMyBucketsResult.java [hunk 1]
```diff
@@ -61,5 +71,13 @@ private Bucket(String name) {
     public String getName() {
       return mName;
     }
+
+    /**
+     * @return the name of the bucket
```
Comment (LuQQiu) [inline core/server/proxy/src/main/java/alluxio/proxy/s3/ListAllMyBucketsResult.java:76]: Update doc?

File: core/server/proxy/src/main/java/alluxio/proxy/s3/ListBucketResult.java [hunk 2]
```diff
@@ -153,6 +173,22 @@ public String getPrefix() {
...
+  /**
+   * @return the delimiter
+   */
+  @JacksonXmlProperty(localName = "Delimiter")
+  public String getDelimiter() {
+    return mDelimiter;
+  }
+
+  /**
+   * @return the prefix
```
Comment (LuQQiu) [inline core/server/proxy/src/main/java/alluxio/proxy/s3/ListBucketResult.java:185]: Update the doc?

File: core/server/proxy/src/main/java/alluxio/proxy/s3/S3RestServiceHandler.java [hunk 3]
```diff
@@ -153,32 +162,30 @@ private FileSystem getFileSystem(String authorization) {
   }
 
   /**
-   * @summary lists all buckets owned by you
+   * Lists all buckets owned by you
```
Comment (LuQQiu) [inline core/server/proxy/src/main/java/alluxio/proxy/s3/S3RestServiceHandler.java:165]: add a .?

**Review by LuQQiu — COMMENTED** [review]
I don't have much knowledge of the PROXY S3 API and nor do the other people in the team I think lol
Did the proxy S3 work well before your changes or it already broke?

File: core/server/proxy/src/main/java/alluxio/proxy/s3/ListAllMyBucketsResult.java (hunk 1 above)
Comment (ZacBlanco) [reply core/server/proxy/src/main/java/alluxio/proxy/s3/ListAllMyBucketsResult.java]: fixed

File: core/server/proxy/src/main/java/alluxio/proxy/s3/ListBucketResult.java (hunk 2 above)
Comment (ZacBlanco) [reply core/server/proxy/src/main/java/alluxio/proxy/s3/ListBucketResult.java]: fixed

**Comment by ggezer** [comment]
@ZacBlanco Can/Should backport these? Is there any dependency on AWS SDK that's not present on older branches?

**Comment by ZacBlanco** [comment]
They probably don't need to be backported unless someone specifically needs these all applied. I mentioned that this potentially breaks backcompat for some responses formats that we previously returned. I personally don't think it is a great idea.

Otherwise though, there's nothing stopping us from backporting.

**Comment by ZacBlanco** [comment]
> I don't have much knowledge of the PROXY S3 API and nor do the other people in the team I think lol
> Did the proxy S3 work well before your changes or it already broke?

The API "worked" but didn't conform to the responses that the latest AWS CLI was expecting, so it would error

**Review by ggezer — APPROVED** [review]

**Comment by ggezer** [comment]
alluxio-bot, merge this please

**Comment by Xenorith** [comment]
alluxio-bot, cherry-pick this to branch-2.6 please

**Pull request merged**